File: Command/Command.CS/Command/Command.cs
using CommandPattern.Executors;
namespace CommandPattern;


public interface Command
{
    void Execute();
    void Undo();
}

public class NullCommand : Command
{
    public NullCommand(){}
    public void Execute(){}
    public void Undo(){}
}

public class LightOnCommand : Command
{
    public Light Light_;

    public LightOnCommand(Light light)
    {
        this.Light_ = light;
    }

    public void Execute()
    {
        this.Light_.On();
    }

    public void Undo()
    {
        this.Light_.Off();
    }
}

public class LightOffCommand : Command
{
    public Light Light_;

    public LightOffCommand(Light light)
    {
        this.Light_ = light;
    }

    public void Execute()
    {
        this.Light_.Off();
    }

    public void Undo()
    {
        this.Light_.On();
    }
}

public class GarageDoorUpCommand : Command
{
    public GarageDoor Door;

    public GarageDoorUpCommand(GarageDoor door)
    {
        this.Door = door;
    }

    public void Execute()
    {
        this.Door.Up();
    }

    public void Undo()
    {
        this.Door.Down();
    }
}

public class GarageDoorDownCommand : Command
{
    public GarageDoor Door;

    public GarageDoorDownCommand(GarageDoor door)
    {
        this.Door = door;
    }

    public void Execute()
    {
        this.Door.Down();
    }

    public void Undo()
    {
        this.Door.Up();
    }
}

public class StereoOnWithCDCommand : Command
{
    public Stereo Stereo_;

    public StereoOnWithCDCommand(Stereo stereo)
    {
        this.Stereo_ = stereo;
    }

    public void Execute()
    {
        this.Stereo_.On();
        this.Stereo_.SetCd();
        this.Stereo_.SetVolume(11);
    }

    public void Undo()
    {
        this.Stereo_.Off();
    }
}

public class StereoOffCommand : Command
{
    public Stereo Stereo_;

    public StereoOffCommand(Stereo stereo)
    {
        this.Stereo_ = stereo;
    }

    public void Execute()
    {
        this.Stereo_.Off();
    }

    public void Undo()
    {
        this.Stereo_.On();
    }
}

public abstract class CeilingFanCommand : Command
{
    public CeilingFan Fan;
    public CeilingFan.Speeds PreviousSpeed { get; set; }

    public CeilingFanCommand(CeilingFan fan)
    {
        this.Fan = fan;
    }

    public abstract void Execute();

    public void Undo()
    {
        switch (this.PreviousSpeed)
        {
            case CeilingFan.Speeds.HIGH:
                this.Fan.High();
                break;
            case CeilingFan.Speeds.MEDIUM:
                this.Fan.Medium();
                break;
            case CeilingFan.Speeds.LOW:
                this.Fan.Low();
                break;
            case CeilingFan.Speeds.OFF:
                this.Fan.Off();
                break;
        }
    }
}

public class CeilingFanHighCommand : CeilingFanCommand
{
    public CeilingFanHighCommand(CeilingFan fan): base(fan){}
    public override void Execute()
    {
        this.PreviousSpeed = this.Fan.Speed;
        this.Fan.High();
    }
}

public class CeilingFanMediumCommand : CeilingFanCommand
{
    public CeilingFanMediumCommand(CeilingFan fan): base(fan){}
    public override void Execute()
    {
        this.PreviousSpeed = this.Fan.Speed;
        this.Fan.Medium();
    }
}

public class CeilingFanLowCommand : CeilingFanCommand
{
    public CeilingFanLowCommand(CeilingFan fan): base(fan){}
    public override void Execute()
    {
        this.PreviousSpeed = this.Fan.Speed;
        this.Fan.Low();
    }
}

public class CeilingFanOffCommand : CeilingFanCommand
{
    public CeilingFanOffCommand(CeilingFan fan): base(fan){}
    public override void Execute()
    {
        this.PreviousSpeed = this.Fan.Speed;
        this.Fan.Off();
    }
}

public class MacroCommand : Command
{
    public Command[] Commands;

    public MacroCommand(Command[] commands)
    {
        this.Commands = commands;
    }

    public void Execute()
    {
        foreach (Command command in this.Commands)
        {
            command.Execute();
        }
    }

    public void Undo()
    {
        foreach (Command command in this.Commands)
        {
            command.Undo();
        }
    }
}

File: Command/Command.Py/src/command.py
from typing import (
    Any, Callable
)


from src.executors import (
    Fan,
    Light,
    GarageDoor,
    Stereo
)


Action = Callable[[bool], None]
Command = Callable[[Any], Action]


def null_command() -> Action:
    def execute(undo: bool) -> None:
        return

    execute.__name__ = "null_command"
    return execute

def light_on_command(light: Light) -> Action:
    def execute(undo: bool=False) -> None:
        if not undo:
            light.on()

        else:
            light.off()
    execute.__name__ = "light_on_command"

    return execute

def light_off_command(light: Light) -> None:
    def execute(undo: bool=False) -> None:
        if not undo:
            light.off()

        else:
            light.on()

    execute.__name__ = "light_off_command"
    return execute


def _fan_undo(previous_speed: Fan.Speeds, fan: Fan) -> None:
    match previous_speed:
        case Fan.Speeds.HIGH:
            fan.high()

        case Fan.Speeds.MEDIUM:
            fan.medium()

        case Fan.Speeds.LOW:
            fan.low()

        case Fan.Speeds.OFF:
            fan.off()

def fan_high_command(fan: Fan) -> None:
    previous_speed: Fan.Speeds

    def execute(undo: bool=False) -> None:
        nonlocal previous_speed

        if not undo:
            previous_speed = fan.speed
            fan.high()

        else:
            _fan_undo(previous_speed, fan)

    execute.__name__ = "fan_high_command"
    return execute

def fan_medium_command(fan: Fan) -> None:
    previous_speed: Fan.Speeds

    def execute(undo: bool=False) -> None:
        nonlocal previous_speed

        if not undo:
            previous_speed = fan.speed
            fan.medium()

        else:
            _fan_undo(previous_speed, fan)

    execute.__name__ = "fan_medium_command"
    return execute

def fan_low_command(fan: Fan) -> None:
    previous_speed: Fan.Speeds

    def execute(undo: bool=False) -> None:
        nonlocal previous_speed

        if not undo:
            previous_speed = fan.speed
            fan.low()

        else:
            _fan_undo(previous_speed, fan)

    execute.__name__ = "fan_low_command"
    return execute

def fan_off_command(fan: Fan) -> None:
    previous_speed: Fan.Speeds

    def execute(undo: bool=False) -> None:
        nonlocal previous_speed

        if not undo:
            previous_speed = fan.speed
            fan.off()

        else:
            _fan_undo(previous_speed, fan)

    execute.__name__ = "fan_off_command"
    return execute